### ISSUE TITLE
Mechanism to include additional compliant algorithms to the FIPS module for custom validations

### ISSUE THREAD

**Issue by t8m** [openssl/openssl#10070]
With addition of new KDF algorithms and also with possible changes in list of approved FIPS algorithms there will be algorithms in the default provider that would be eligible for FIPS validation in the FIPS provider for consumers of the FIPS provider who are willing to do their own custom validation.

There should be a simple (build-time) mechanism how to include such algorithms in the FIPS module.

**Comment by paulidale** [comment]
An easier way to pick and choose inclusions in providers would be nice.

**Comment by mattcaswell** [comment]
I envision some text file which lists the algorithms to be included in the FIPS module. Out-of-the-box we just use a default list, but a parameter to Configure could specify a custom one.

Now if only we had a perl and build system guru that could figure out how to do that...

**Comment by paulidale** [comment]
What if the text file were called something like `build.info` and it was placed in its own directory and Configure was told where to look?

**Comment by mattcaswell** [comment]
> What if the text file were called something like build.info and it was placed in its own directory and Configure was told where to look?

Yes, that could work. Although at the moment there are multiple build.info files that control what files get included in fips.so. We'd also have to build the OSSL_ALGORITHM lists in fipsprov.c based on what algorithms are available. That could be with lots of "defines" (one per algorithm) - which sounds a little unwieldy, or possibly through some dynamic build time generation of the code (a template perhaps)

**Comment by mattcaswell** [comment]
Hmmm...perhaps a build.info file might expose a little too much of our internal build mechanism. We'd ideally like the text file to be as simple as possible, so that a third party can easily add or remove algorithms.

**Comment by levitte** [comment]
You're over-thinking it. All you really need to do is to collect all the implementations in a static library, and link each provider with it. Symbol resolution should take care of the rest. So all you have to do is exactly what you do today, i.e. set up `OSSL_ALGORITHM` arrays in each provider main (`providers/default/defltprov.c`, `providers/fips/fips prov.c` and `providers/legacy/legacyprov.c`) and let the magic happen.

This assumes that the algo implementations themselves are provider agnostic, so they don't have or need to have any check of `FIPS_MODE`. Worst case scenario, we may have to have a separate static library for FIPS... I would rather avoid that if we can, or possibly have a side static library that fulfill whatever the FIPS module needs.

We should think of reorganizing a little bit, so *all* implementations are in `common`. One easy reorg would be for each implemented algorithm set to live in its own subdirectory, with its own `build.info`, and then add another `SUBDIRS = name` (it's cumulative) in the `build.info` in the directory above. That's enough to get it included.
And yeah, people who want to "patch" openssl (rather) will have to learn enough `build.info` to copy another and make their mods, and to patch the appropriate file(s) to bind the implementation.

**Comment by levitte** [comment]
> Now if only we had a perl and build system guru that could figure out how to do that...

I wonder who that would be...

**Comment by mattcaswell** [comment]
> So all you have to do is exactly what you do today, i.e. set up OSSL_ALGORITHM arrays in each provider main (providers/default/defltprov.c, providers/fips/fips prov.c and providers/legacy/legacyprov.c) and let the magic happen.

Right - that's half the problem though. We want to be able to dynamically (at build time) specify what's in those arrays for the fips provider.

**Comment by levitte** [comment]
So basically, you want a generated fipsprov.c. That shouldn't be too hard, we do have the tooling for it.

**Comment by levitte** [comment]
Should we generate the other provider mains that way too ?

**Comment by mattcaswell** [comment]
> So basically, you want a generated fipsprov.c.

Yes. And the ability for a third party to be able to modify the list that gets included without having to change any sources.

> Should we generate the other provider mains that way too ?

Hmm. Interesting idea. Not sure its strictly necessary at the moment.

**Comment by levitte** [comment]
I think that the main thing will have to be to figure out is how to tell the template the symbols for the diverse `OSSL_DISPATCH` tables, as well as the names associated with them (those names that, for some reason, shouldn't be located in source together with said dispatch table [ahem])

**Comment by levitte** [comment]
Perhaps some kind of per implementation manifest file?

**Comment by mattcaswell** [comment]
> Perhaps some kind of per implementation manifest file?

This is essentially what I was talking about with the text file I mentioned above. We would have our own "default" manifest file, but there could be a Configure option to specify the location of a different one.

**Comment by levitte** [comment]
I'm pondering that `build.info` could declare manifest files. Do you have a format in mind? Otherwise, they can easily be done in perl, like our config target files, and all they would have to do is to present a small hash table, something like this:

``` perl
(
    {
        type => 'OSSL_ALGORITHM', 
        operation => 'digest', 
        dispatch => 'blargh_functions', 
        names => "BLARGH:id-blargh", 
        properties => "whatever=42"
    }
) 
```

The template will have to collect then all and generate the appropriate algorithm tables.

Manifests could then be attached to their goal provider, like this:

```
MANIFEST[fips]=common/digests/blargh/manifest.txt
```

As usual, such values would be cumulative

**Comment by mattcaswell** [comment]
I had no particular format in mind, except to say the simpler the better. I was expecting a very simple list of algorithm names (i.e. it doesn't have to have the ability to specify the properties, or the aliases, or the functions - all of that would be fixed. We just need to be able to say "I want SHA2-256, AES-128-CBC and AES-128-GCM".

**Comment by levitte** [comment]
So, er, someone with a new KDF (hey, such as KDB5KDF !), and somehow, we need to figure out the name of OSSL_DISPATCH array, what properties it should have apart from "fips=yes" (which we would obviously force on them) and it's name(s)? Just say a name and *assume* that they have named, for example, the dispatch array according to our code? I dunno about you, but I for one prefer a tad more declaration...

**Comment by levitte** [comment]
I assume that what you're trying to say is that instead of manipulating a `build.info` with MANIFEST statements, you just want to be say, in the template, that I want manifest this this and that.  I'm not sure if that makes things easier for everyone to hack the template, but sure...

But the questions we need answering are:

1.  how do you say "I want BLARGH, ..."
2.  how does the template get the information it needs to build up the OSSL_ALGORITHM tables

My answer is currently:

1.  I want manifest "BLARGH-manifest.txt"
2.  The information is in that manifest.

**Comment by t8m** [comment]
If I am adding new algorithms, I can patch things in the existing build.infos and sources. The original intention was limited to what Matt wrote in the previous comment.

**Comment by simo5** [comment]
Indeed the point is just making it possible to select what goes in the "Fips module" from the already available algorithms.
If someone is patching in a completely new algorithm they will also create the default build info for that algorithm.
The name of that new algorithm can then be used in the manifest to pull it into the fips module boundary.

**Comment by mattcaswell** [comment]
> how do you say "I want BLARGH, ..."
A simple list of algorithm names in a text file (or manifest or whatever you want to call it)

> how does the template get the information it needs to build up the OSSL_ALGORITHM tables

The template can have have a perl hash of all the available algorithms which contains all the function names, properties etc. It then has to read the manifest file and generate the OSSL_ALGORITHM tables and spit out a line where it has an entry in its hash which is also in the manifest file.

**Comment by levitte** [comment]
So uhmmmm, let me see if I get this right...  all you want is to be able to add or remove entries from the `OSSL_ALGORITHM` array, but you don't want to do that directly in `fipsprov.c`...  is that it?

**Comment by mattcaswell** [comment]
Right - you should be able to specify at Configure time the list of algorithms that actually end up in fipsprov.c.

**Comment by levitte** [comment]
I forgot about the configurability of it...  Er, sure.  How would like to specify that list?  One option with a comma-separated list of names?  Variants of `no-` / `disable-` options that are all `no-fips-` / `disable-fips-`?  Actually, the latter would be very much along the lines of how we already do things...

**Comment by mattcaswell** [comment]
> I forgot about the configurability of it... Er, sure. How would like to specify that list? One option with a comma-separated list of names? Variants of no- / disable- options that are all no-fips- / disable-fips-? Actually, the latter would be very much along the lines of how we already do things...

Well, I had in mind a file that lists all the algorithm names - and the Configure option would just provide the name of the file. Because there are likely to be a *lot* of algorithm names - so specifying them all individually on the command line seems like a big pain.

**Comment by levitte** [comment]
Ah...  our usual M.O. is to ask users to *disable* things, so ask yourselves, do you expected to *disable* many of them?  @simo5, I assume you have some kind of idea here.

**Comment by mattcaswell** [comment]
> Ah... our usual M.O. is to ask users to disable things, so ask yourselves, do you expected to disable many of them? @simo5, I assume you have some kind of idea here.

My expectation would be yes. It's quite possible (even likely I would say) that someone just wants a very minimal FIPS module with a handful of algorithms in it just suitable for their own purpose.

**Comment by simo5** [comment]
@levitte in this case I agree with @mattcaswell that a file would be easier to manage for a few reasons:
1) it is easy to store the file as a downstream patch/collateral for future reference.
2) it is easy to take the upstream file as an initial starting point and just add/remove things as needed from there.
3) it is easy to diff the upstream and downstream files to see the differences (obvious if this is a downstream patch, but in general as well).

As for disabling it really depends on what you are doing, but I can point you to a fitting use case that would require disabling *a lot* of stuff.
We use openssl embedded into a shim bootloader for secure boot verification. Because the binary format is special we cannot use the normal module we certify in here so we currently do not FIPS validate this component.
But if we were to certify it (and there is a chance) we would not want to spend a lot of money to validate all algorithms (NIST is now moving to a fee structure that depends on the number of algorithms tested for CAVS/ACVP), therefore we would probably want to disable fundamentally all algorithms except RSA and ECC signature verification.
If that required a long list of -disable configure options it would become painful to deal with it quite fast as we'd have to track at each rebase whether new algorithms were added and have to be disabled and so on.

**Comment by simo5** [comment]
Additionally if you want to reuse this facility to create multiple modules, maybe you have FIPS but you also have some special acceleration engine, or pkcs#11 module, or a new certification form a different country, or ... then configuration options would explode quite rapidly and would be impossible to deal with.
I am not saying you should allow to build multiple custom modules from the get-go, but if that's something that may happen in future, better set up things so it won't be hard/disruptive I think.

**Comment by levitte** [comment]
Okie

**Comment by levitte** [comment]
So, it sounds like the plan is this:

- Make it possible for Configure to handle a per-provider file of names (currently limited to the FIPS provider).  The exact meaning of those names is left to the building code to deal with (we might even just take the file name and let the build code deal with the contents)
- Move all of `providers/default/{ciphers,digests,kdfs,macs}` to `providers/common/` (and `providers/default/include/internal/blake.h` and `providers/common/include/internal/`)
- Create an internal static library that includes *all* of the object files from `providers/common/**`
- Generate `providers/fips/fipsprov.c` with the list of names that Configure got, coupled with a database of facts associated with each name.

Did I miss something?

**Comment by mattcaswell** [comment]
> Generate providers/fips/fipsprov.c with the list of names that Configure got, coupled with a database of facts associated with each name.

If Configure didn't get a list of names we use a default list.

**Comment by levitte** [comment]
> If Configure didn't get a list of names we use a default list.

So everything from the database of facts, basically :wink:
Yeah, that was my assumption.

**Comment by levitte** [comment]
Ah, I realised that we can't use this method for anything that goes into libcrypto, such as the default provider.  The reason is that slurping one static library into another is a hard thing to do...  and not really worth the bother in this case.

**Comment by petrovr** [comment]
Richard Levitte wrote:
> You're over-thinking it. All you really need to do is to collect all the implementations in a static library, and link each provider with it. Symbol resolution should take care of the rest. So all you have to do is exactly what you do today, i.e. set up `OSSL_ALGORITHM` arrays in each provider main (`providers/default/defltprov.c`, `providers/fips/fips prov.c` and `providers/legacy/legacyprov.c`) and let the magic happen.
Before (FIPS module 1.2 or 2.0) management of fips algorithms was more 
at run time.

Configuration at build time only is functional degradation.




> This assumes that the algo implementations themselves are provider agnostic, so they don't have or need to have any check of `FIPS_MODE`. Worst case scenario, we may have to have a separate static library for FIPS... I would rather avoid that if we can, or possibly have a side static library that fulfill whatever the FIPS module needs.

No there must be check at run time if fips is activated or not. Let see 
e_nss engine. It marks RSA/DSA/EC algorithms as available in FIPS mode. 
At same time if OpenSSL is run in FIPS mode but NSS module is not engine 
fails to initialize.

>
> We should think of reorganizing a little bit, so *all* implementations are in `common`. One easy reorg would be for each implemented algorithm set to live in its own subdirectory, with its own `build.info`, and then add another `SUBDIRS = name` (it's cumulative) in the `build.info` in the directory above. That's enough to get it included.

In additional PKIX-SSH is another sample.
Point here is that some secure shell algorithms are not allowed if 
crypto runs in FIPS mode. Using them lead to program termination. This 
is not acceptable for secure program.
So program must not at run time what is acceptable (chipper, mac) and do 
not announce or avoid to use them if fips is activated.

By design PKIX-SSH works with FIPS validated modules from different 
vendors and different version of cryptographic library.
In such case static configuration is nightmare. This is reason to use 
"dynamic" checks.

> And yeah, people who want to "patch" openssl (rather) will have to learn enough `build.info` to copy another and make their mods, and to patch the appropriate file(s) to bind the implementation.

It will be good to extend at run time.
Usually secure token fulfil fips requirement and so in PKIX-SSH pkcs11 
support is marked as available in fips mode.


I would like to be sure that new FIPS 3.0 support will allow dynamic 
enhancement.


Regards,
Roumen Petrov

**Comment by levitte** [comment]
> Before (FIPS module 1.2 or 2.0) management of fips algorithms was more at run time.

This discussion is what about gets built into the FIPS module.  That can't be managed at run-time, not in 1.2, not in 2.0 and not now.

**Comment by levitte** [comment]
> I would like to be sure that new FIPS 3.0 support will allow dynamic enhancement.

This sounds like a subject for another issue.

**Comment by mattcaswell** [comment]
> I would like to be sure that new FIPS 3.0 support will allow dynamic enhancement.

Anyone can provide a FIPS provider with additional algorithms in it that aren't in the main OpenSSL FIPS module. They can have the "fips=yes" property and libcrypto will treat them in the same way as algorithms from *our* FIPS module. So in this way you will be able to do this.

 But you wouldn't (under then proposals discussed in this issue) be able to dynamically add non-OpenSSL algorithms into the OpenSSL FIPS module.

**Comment by mattcaswell** [comment]
> So everything from the database of facts, basically wink
> Yeah, that was my assumption.

Well...not necessarily. Firstly it depends on whether the database of facts *only* includes FIPSable algorithms. Obviously we don't want non-FIPSable algorithms in the FIPS module. Secondly this whole requirement came from #9949 which proposes to add a FIPSable algorithm that we *won't* be adding into our own FIPS module (well probably not anyway). That's an algorithm that wasn't in our original scope and don't want to add into the lab scope, but could be included in third party's FIPS modules.

**Comment by simo5** [comment]
The only thing that really make sense to do dynamically is to restrict the algorithms available in FIPS mode to a subset of the compiled in algorithms.
Adding algorithms makes no sense because they wouldn't be validated algorithms in the first place so a flat violation of what FIPS means in its essence.

For restricting algorithms we use crypto-policies in Fedora/RHEL, and i is generally a good thing to be able to further restrict at runtime the default set of algorithms. Applications should always be able to override such a restrictions though.
In the FIPS case using an algorithm that is no in the FIPS module should be permitted only if the application explicitly states that such use will not be for cryptographic or user data protection purposes.

My 2c.

[none of this matter for the purpose of this issue though, which is about what gets compiled in the FIPS module, not how to access functions at runtime]

**Comment by levitte** [comment]
> Firstly it depends on whether the database of facts _only_ includes FIPSable algorithms.

Each implementation could have tags associated with them, saying where they are expected by default, and where they are allowed.

**Comment by paulidale** [comment]
What about something along the lines of the Linux kernel's `make config` (or better `make menuconfig` or best `make xconfig`).  Well the kernel's stuff as it is now since that was ten years ago...

**Comment by levitte** [comment]
It's been too long since I built my own kernel, so I no longer have the faintest clue what `make config` does.

**Comment by paulidale** [comment]
It asks an annoyingly long series of Y/N questions to customise the kernel.  Answers are pre-populated generally.

**Comment by levitte** [comment]
Oh, *that*!

**Comment by levitte** [comment]
I'm not so keen on that idea...

**Comment by richsalz** [comment]
OMG, please no!  A text file, and a config option to point to a different text file.  This has to be portable to a bunch of platforms.  It is quite okay to sanity-check the config file (find-fips-nits anyone?) and fail the build if *the FIPS module developer* makes errors.  This is not the kind of thing that the general public will need to edit or modify.

Please, no UI interaction.  We need a text file anyway, for automation.

**Comment by paulidale** [comment]
> All you really need to do is to collect all the implementations in a static library, and link each provider with it.

I seem to remember suggesting exactly this :)

It means building everything twice -- once with FIPS and once without.

**Comment by paulidale** [comment]
Also worthy of consideration is that any new algorithms (KDFs in this case), cannot be FIPS validated even though they can be included within the FIPS boundary.  They can be vendor affirmed once NIST recognises their existence (okay too strong) and they can be included on the basis of being not security realated or for legacy purposes.

Let's not be hasty in creating a disaster before we have a viable FIPS provider.  Fine tuning like suggested here is great, but we've a lot to do before it becomes viable.  We should always remain cognisant of the future possibility however.

**Comment by levitte** [comment]
> > All you really need to do is to collect all the implementations in a static library, and link each provider with it.
> 
> I seem to remember suggesting exactly this :)
> 
> It means building **everything** twice -- once with FIPS and once without.

Depends...  we could have one static library that's *only* the algo implementation object files.  I've looked around, and only found two places where `FIPS_MODE` is checked, and it's to set some constants, which could be relegated to separate object files (make them non-static constants, and refer to them from the implementation), one for FIPS and one for non-FIPS.  All object files that are just support (including everything under `crypto/`) can be built as they are now.

This means we could end up getting *less* double building.

**Comment by t8m** [comment]
> Also worthy of consideration is that any new algorithms (KDFs in this case), cannot be FIPS validated even though they can be included within the FIPS boundary. They can be vendor affirmed once NIST recognises their existence (okay too strong) and they can be included on the basis of being not security realated or for legacy purposes.

I do not really understand this - the KDFs we are adding are FIPS approved algorithms and if we do downstream our own full FIPS module validation we can have them on the same level of FIPS validation as other algorithms in the FIPS module.

**Comment by levitte** [comment]
@t8m, if you look at the bottom of https://www.openssl.org/docs/OpenSSL300Design.html, you will see what algorithms were anticipated.  Those are *definitely* going to be part of the *source code validation* of the FIPS module.  Anything added to that list is what I believe others have called "FIPS-able", and if I understand correctly, it's debated if they can be part of the *source code validation* or not.  However, that should have no impact on any validation of a binary module.

I hope that this makes the distinction clearer...  In other words, there shouldn't really be any difference between what we call "FIPS" and what we call "FIPS-able" for building the module.  They should all go in by default.

**Comment by paulidale** [comment]
@t8m, I over generalised (as usual).  My understanding is that not _all_ of the new KDFs can be validated.  SP 800-135 lists those additional algorithms which can be: SSHKDF is there but I don't see, for example, the Kerberos KDF (RFC 3961) on the list.

Of course this doesn't mean that they can't be included inside the FIPS boundary.

**Comment by t8m** [comment]
OK, now I get it. And yes, RFC 3961 KDF is something that is not explicitly approved algorithm but our lab told us it is in compliance with SP800-133, so it can be used.

**Comment by levitte** [comment]
I've started coding build stuff that should help.  [ahem] assume magic (in the words of Arthur C. Clarke)

**Comment by levitte** [comment]
You might want to have a look at #10088 and approve or complain, and at #10124, where I just started the algorithm selection work

**Comment by simo5** [comment]
Question, if an application that otherwise wants to be FIPS compliant has an exception use for a non-approved algorithm, how does it access it ?

Say, something like the use of a signature verification function for an algorithm that is not approved anymore just for compatibility reason, maybe to be able to re-sign with a newer algorithm just after verification.

Also that function may be implemented with a key on an HSM, so a pkcs11 module would have to be used, is there a way this can be configured by an admin without changing the application?
Or would it need to be a known valid configuration at the time the application is build and provided for by the specific application configuration mechanisms ?

**Comment by levitte** [comment]
That's a matter of having the necessary providers loaded and explicit fetching with properties that select the desired implementation.

Loading providers can be done through configuration.

**Comment by simo5** [comment]
I am trying to understand if these are actions that need to be coded explicitly in the application or can be controlled externally ?

**Comment by levitte** [comment]
The following can be controlled with the OpenSSL config file:

-   loading providers
-   activating providers
-   specify default properties

This is quite powerful, and as long as each algorithm is only implemented by one provider, that should be enough to use pre-3.0 code for any algorithm that was available in pre-3.0 libcrypto, if the appropriate providers are loaded. 

If there is more than one implementation of any given algorithm with the available properties (default ones unless an explicit fetch is performed), I *think* you simply get one of them. If you want to be in better control, you must do an explicit fetch, and that will require changes to the application code.

Did this help?

**Comment by levitte** [comment]
If I understand what you write correctly yes.

To be precise, you might have the default OpenSSL configuration file, which is `/usr/local/ssl/openssl.cnf` in a default install directly from our source, or `/etc/ssl/openssl.cnf` on most Debian derivations, and you may also have an application specific configuration in a separate so called OpenSSL config file.  The latter matches your second point insofar that there are knobs for the admin to twist.

**Comment by t8m** [comment]
Let's make the question more concrete:
1. We have the global config file specifying fips=yes.
2. We want to be able to use non-fips algorithm for legacy non-security-relevant purpose.
How can the application fetch the non-fips algorithm implementation to override the config file fips=yes property? We do not want to explicitly use any particular provider as the algorithm might be in default or legacy provider or maybe even in a third party pkcs11 provider.

**Comment by levitte** [comment]
The easiest is to use a separate config file specific to the application, with exactly the setup you want. The environment variable `OPENSSL_CONF` can be used to tell libcrypto where that config file is.

**Comment by t8m** [comment]
The situation is that we have a full control of the application code but we do not want all the provider implementation choices be affected by this "allow non-approved" override. This should be specific to a particular OpenSSL API call only - such as use of MD5 in file name normalization or similar non-security relevant things. For everything else the default config file should apply - if it specifies fips=yes, fips provider should be used.

**Comment by simo5** [comment]
> The easiest is to use a separate config file specific to the application, with exactly the setup you want. The environment variable `OPENSSL_CONF` can be used to tell libcrypto where that config file is.

Sounds impractical if you have many applications. The problem is that we know a specific algorithm is non-fips relevant at application creation time, so we can mark it somehow, what we do not know is *when* the application is used in FIPS mode, because that is a deployment choice.
So we do not have a good way to handle "fips enablement" with OPENSSL_CONF because we do not know when to "conditionally" apply it.

Unless OPENSSL_CONF can be somehow additive and we only specify configuration for the specific algorithm. Even so it is suboptimal because within an application there may be both compliant and non-compliant uses of an algorithm. So the indication that the fips module should be ignored is rally a property of the specific call site not a general property to associate to an algorithm across the whole application (this is even more true when we start considering libraries that are pulled into an application and may be using algorithms in ways that differ from the rest of the application.

Currently we have "NON_FIPS_ALLOW" flags that have a reasonable behaviour (localized, easy to use, understandable), we are looking into how that behaviour can be retained in the v3.0 world.

**Comment by levitte** [comment]
That sounds like it's beyond the capabilities of our config syntax, even. Sounds like you would like an embedded scripting language, so err, Lua in version 5?
(no, Lua or any embedded scripting language is *not* a serious suggestion) 

I would say that, at the very least now, what you're talking about would require application changes.

**Comment by levitte** [comment]
Btw, what's been discussed here today seems quite a bit off-topic for this issue...

**Comment by simo5** [comment]
@levitte I know this is somewhat specific, but we do not need a programming language to configure stuff, just an indication that a call can skip the FIPS module and use whatever elase is configured on the system.
In this case the FIPs module is the special case because (afaik) it is supposed to stop you from using non-approved algorithms.
Or am I wrong in thinking that a vr openssl  configured with fips=yes will prevent things like RC4 from being used ?

**Comment by simo5** [comment]
> Btw, what's been discussed here today seems quite a bit off-topic for this issue...

Sorry, initially it wasn't clear to me is the algorithm selection had a bearing on this, feel free to push me to a better place if more discussion on this topic is needed.

**Comment by levitte** [comment]
> Or am I wrong in thinking that a vr openssl configured with fips=yes will prevent things like RC4 from being used ?

No, that's entirely correct, since there will not be an RC4 implementation with that property.

**Comment by t8m** [comment]
I've opened #10129 for the override mechanism API call.

**Comment by mattcaswell** [comment]
>> Or am I wrong in thinking that a vr openssl configured with fips=yes will prevent things like RC4 from being used ?
>
> No, that's entirely correct, since there will not be an RC4 implementation with that property.

Unless an application uses explicit fetch and overrides the default properties. It is not necessary for an application to have a separate config file for this. It just needs to be explicitly coded to override the default properties.

**Comment by levitte** [comment]
> It just needs to be explicitly coded

I think they're asking how they can avoid that.  #10129 does bring clarity

**Comment by mattcaswell** [comment]
> I think they're asking how they can avoid that. #10129 does bring clarity

Oh really? That wasn't my interpretation of this request or #10129 at all.

**Comment by simo5** [comment]
@levitte just for clarity, I was basically looking to see if the NON_FIPS_ALLOW flag could be maintained to avoid changing that in all applications that makes use of it once we transition to 3.0.

Applications need to be explicit, there is no way around it because it's only specific uses at specific call sites that can be exempt, not an algorithm in general.

I think #10129 is the right place to discuss though, so deferring to there.

**Comment by paulidale** [comment]
> If there is more than one implementation of any given algorithm with the available properties (default ones unless an explicit fetch is performed), I _think_ you simply get one of them.

You will get exactly one of them and there is no guarantee which since we reserve the right to change the ordering at anytime.  Properties are the way to pick a specific algorithm.

There is a call to query for all implementations, again the order is deliberately undefined.  I'm not sure if we expose it.

**Comment by mattcaswell** [comment]
This feature won't be in beta 1. Taking it out of that milestone.

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
